### ISSUE TITLE
TCK: Replace short and long delays with calls to LRARecoveryService 

### ISSUE THREAD

**Issue by xstefank** [eclipse/microprofile-lra#249]
Long delays are present because of repeated recovery calls. This can now be replaced by calls to LraRecoveryService.

The use of short delays in the TCK should also be reviewed with the aim of providing a less brittle mechanism.

**Comment by mmusgrov** [comment]
What about the short delays?

**Comment by mmusgrov** [comment]
Implementations needs to be able to decide why they are being asked to delay/retry and the interface should reflect the different reasons why. So something like:

    boolean waitForCallbacks(); // returns after an attempt to notify all participants was made
    boolean retryCallbacks(); // some callbacks previously failed so request

or something else that captures the same information.

**Comment by xstefank** [comment]
@mmusgrov short delays are tracked in #250

**Comment by mmusgrov** [comment]
The two issues are similar (and either PR will need to cover both issues):

1. short delay is used to wait for participant callbacks to be sent.
2. long delay is used in crash recovery scenarios for the callbacks to be redelivered after a previous attempt failed.

The two features could be combined into a single feature, eg:

``` java
(1) waitForFirstGroupOfCallbacks() // the successful path where participants behave
(2) waitForNextGroupOfCallbacks() // the failure path where things eventually become consistent
(3) waitForLastCallback() // the after notifications have been received
```

In non recovery scenarios the test would call (1) and would call (2) in recovery scenarios. (3) is currently not useful.

So if we discuss the two scenarios as a single feature then we can avoid having to cross reference comments between the two issues.

**Comment by xstefank** [comment]
fair enough, if we all agree that all timeouts will be removed and eventual consistency will be handled with this push model. However, long delays should still be replaced with LraRecoveryService which I will do first before the rewrite of all TCK tests for this. I would like to keep this issue open for the removal of long delays and then do short delays in separate PR.

**Comment by mmusgrov** [comment]
waitForNextGroupOfCallbacks() would be the equivalent long delay (and would go in your LraRecoveryService API)
waitForFirstGroupOfCallbacks would be the equivalent of short delay (which similarly would go in your LraRecoveryService API)

Why can't you do both of those in a single issue and PR since any discussions we have about waiting are best had in a single place.

**Comment by xstefank** [comment]
`waitForNextGroupOfCallbacks()` would then had to wait for the recovery if there are no calls to `LraRecoveryResource#triggerRecovery()` -- this is what I want to start with. The subsequent rewrite will be going from `LRA action -> short delay -> assert` to `LRA action -> waitForCallbacks`.

**Comment by mmusgrov** [comment]
> `waitForNextGroupOfCallbacks()` would then had to wait for the recovery if there are no calls to `LraRecoveryResource#triggerRecovery()` -- this is what I want to start with. The subsequent rewrite will be going from `LRA action -> short delay -> assert` to `LRA action -> waitForCallbacks`.

waitForNextGroupOfCallbacks is the equivalent to triggerRecovery.

Or we could just have a single method:

```java
waitForCallbacks(); // wait until each participant has been contacted
```

Then calling the same method a second time would be equivalent to triggerRecovery/longDelay;

**Comment by xstefank** [comment]
Sorry, I misunderstood you originally. You mean that these methods will be implemented by the implementation not in the TCK itself which is something I planned to do. I would prefer to have a separate service in the TCK on which you can call `waitFor(CallbackBuilder.compensate(1).afterLRA(1))` or something similar and this would be checked in the TCK itself. So implementation will still need to implement only `LraRecoveryService`. WDYT?

**Comment by mmusgrov** [comment]
I am discussing what the API for the LRARecoveryService interface (that you defined) should look like (the Narayana code base would need to do the actual logic implied by waitForCallbacks). So the TCK would provide something similar to the following:

```java
package org.eclipse.microprofile.lra.tck.service.spi;

public interface LRARecoveryService {
   boolean waitForCallbacks(); // wait until each participant has been contacted
}
```

**Comment by xstefank** [comment]
Yes, as I said, we had a misunderstanding but now I can see your point. And I suggested an alternative. What are you thoughts on the alternative, please?

**Comment by mmusgrov** [comment]
> Yes, as I said, we had a misunderstanding but now I can see your point. And I suggested an alternative. What are you thoughts on the alternative, please?

I did not understand your alternative. Can you rephrase it please?

**Comment by xstefank** [comment]
I think rewriting LraRecoveryService to the interface you proposed would introduce a lot of repeated code in the individual implementations (for instance, the implementation needs to track whether the repeated call is needed (aka removing triggerRecovery)) so I would like to rewrite TCK to a pooling model. This would mean we can have a single global timeout config value `lra.tck.timeout` which would only define the top limit on the waits and individual tests would look like this:

```java
lraExecutionService.waitFor(CallbackBuilder
  .compensate(1, LraResource.class).build()); // this will return only when Compensate is called in LraResource class or it will throw an exception / fail the test if lra.tck.timeout elapsed

// now we know that Compensate was called so no need for asserts or maybe just verify that Complete wasn't called

lraRecoveryResource.triggerRecovery(); // possibly this can be now non-blocking

lraExecutionService.waitFor(CallbackBuilder)
  .status(1, LraResource.class)
  .afterLRA(1, LraResource.class)
  .build());
```

I forgot to mention lraExecutionService would definitely integrate with LRAMetricService to get this done. But the idea is that the tests will be looking on the TCK executions correctly from the point of users who don't know when the callbacks are called.

**Comment by mmusgrov** [comment]
I think that is over complicating the requirement. We just need to search and replace:

 1) applyShortConsistencyDelay` with lraRecoveryService.waitForCallbacks
 2) applyLongConsistencyDelay` with lraRecoveryService.waitForCallbacks; lraRecoveryResource.waitForCallbacks

ie there should be no timeouts (unless we want to catch a hanging test)

**Comment by xstefank** [comment]
I agree it would be more complicated but it would be more closely related to external view of the users. Maybe later not for 1.0.

> applyLongConsistencyDelay` with lraRecoveryService.waitForCallbacks; lraRecoveryResource.waitForCallbacks

if there is no `triggerRecovery()` between these calls how will the implementation know that it now have to wait for new callbacks? How many callback should we wait for before returning? So if there is Compensate and AfterLRA how does implementation know that it shouldn't return after Compensate?

**Comment by rdebusscher** [comment]
Not a fan of the long builder style either, but it is very concise.

But the long wait is not just waiting twice for callback. If I remember correctly, the long and short wait are there for 2 different types of 'actions/processes' which needs to be completed.

So they should be different methods so that TCK implementors can do the correct thing in both cases.

**Comment by xstefank** [comment]
So ok, after reevaluating and checking the applications of individual delays, here is my summary:

- short delay is for waiting for eventual delivery of callbacks (e.g. TCK calls cancel, after short delay all Compensate callbacks should have been called)
- long delay is for waiting for periodic recovery to replay Status or Complete/Compensate calls(e.g. TckUnknownStatusTests#compensateRetry()) 

What I can see as a solution integrating Mike's ideas:
- replace short delay with a new method `waitForCallbacks()` which will be implemented by the implementation
- since `LraRecoveryService#triggerRecovery` is already waiting for Status or Complete/Compensate calls to be redelivered, only replace all long delays with calls to `triggerRecovery()`

Can I implement this?

**Comment by rdebusscher** [comment]
+1

**Comment by mmusgrov** [comment]
+1

**Comment by mmusgrov** [comment]
waitForCallbacks and waitForRecovery looks more harmonious.

Does waitForCallbacks return when the implementation has issued the callbacks or does it wait to see what kind of response was returned?

Does triggerRecovery/waitForRecovery wait until all participants have reached one of our terminal states (and if so what is the mechanism to avoid a hang)? Or does it wait until the implementation has re-issued the callbacks to any participant that failed on the initial attempt? Or does it do something additional/different from those two alternatives.

What return types do you plan to use and what will their meaning be?

**Comment by xstefank** [comment]
> waitForCallbacks and waitForRecovery looks more harmonious.
ok, I can rename triggerRecovery

> Does waitForCallbacks return when the implementation has issued the callbacks or does it wait to see what kind of response was returned?

I presume void should be enough. The implementation will guarantee that when this method returns all callbacks have been processed.
 
> Does triggerRecovery/waitForRecovery wait until all participants have reached one of our terminal states (and if so what is the mechanism to avoid a hang)? Or does it wait until the implementation has re-issued the callbacks to any participant that failed on the initial attempt? Or does it do something additional/different from those two alternatives.

It should reissue callbacks. When the method returns the TCK will check the results. So implementation must make sure to not return before validation can happen. Hanging is also only possible on the implementation side so I don't think it's necessary to track it in TCK. If implementation hangs it is its responsibility to take care of this situation.

> What return types do you plan to use and what will their meaning be?

void should be enough. Do you have any reasons why we would introduce something else? If we are moving the responsibility to the implementation let's rather do fewer things in the TCK because then it can get messy.
